### DISCUSSION TITLE
Cannot flash or compile Hello world on nrf52840

### DISCUSSION THREAD

**Discussion by emaddos** [contiki-ng/contiki-ng#1794] [Q&A]
Hello, 

I have a project with IoT devices that have chip-set nrf528420. I installed contiki in both Ubuntu 20.04 and 18.04 but I cannot compile using the IoT Device.
Did you test contiki on a device with the same chip-set and can you guide me to make it work.?

Thanks

**Comment by pjonsson** [comment]
About a month ago, I followed the instructions on the wiki (https://github.com/contiki-ng/contiki-ng/wiki/Platform-nrf52840) and things worked for me on Ubuntu 20.04.

**Comment by alexstanoev** [comment]
Could you share the output of `make TARGET=nrf52840` for hello-world?

**Comment by emaddos** [comment]
@pjonsson What is the device you were working with

**Comment by mbaddeley** [comment]
Maybe a stupid question, but you are using Contiki-NG right? not the old Contiki OS?

**Comment by emaddos** [comment]
@MBradbury Yes I am using Contiki-NG

**Comment by emaddos** [comment]
dos@dos-PC:~/contiki-ng/examples/hello-world$ make TARGET=nrf52840 hello-world.upload
nrfjprog -f nrf52  --sectorerase --program build/nrf52840/dk/hello-world.hex
ERROR: No debuggers were discovered.
NOTE: For additional output, try running again with logging enabled (--log).
NOTE: Any generated log error messages will be displayed.
make: *** [../../arch/cpu/nrf52840/Makefile.nrf52840:127: hello-world.upload] Error 41

**Comment by pjonsson** [comment]
I'm not sure what my device is called, but it's the one that uses TARGET=nrf52840 BOARD=dongle.

This is a recollection of what I had to do rather than a reaction to the error message from your make: unfortunately I don't recall if it was the nrf5340 or the nrf52840 device, but one of them required me to run the flash utility with some other parameters before "make hello-world.upload" started succeeding. That command wasn't described on the Wiki page I linked earlier.

Unless you want to wait for Alex to chime in, I would try adding "--log" to the nrfjprog invocation to get more information like the error message suggests.

**Reply by rahav** [reply]
I am learning that Nordic motes activate a mechanism on the mote called `readback protection` after each time they are programmed. So far I did not fine a way to disable this, maybe @alexstanoev can chime in on this. Therefore, before programming the following command is required:

```nrfjprog --recover```

While understand plausible need for the `readback protection`  feature (security, it prevents malicious actors from reading the code on the mote), I do not think it belongs on motes that are use for experimental purposes. We program these motes again and again, and pretty soon `nrfjprog --recover` becomes annoying.

Suggestions:
- Find a way to disable readback protection on the motes themselves (a jumper or setting)
- Add `nrfjprog --recover` or a lighter command (I don't know if this command does too much and what is the impact to the flash) to the compilation (`make`) process
- We certainly need to add this "feature" to the docs

**Reply by alexstanoev** [reply]
Access port protection does work differently on newer revision devices as it needs to be disabled by both hardware (write to UICR once) and software (write to NRF_APPROTECT->DISABLE) after every reset.

The MDK used in the old nrf52840 port is too old for this, but if you use the nrf platform you can disable APPROTECT like this:

```C
if ((NRF_UICR->APPROTECT & UICR_APPROTECT_PALL_Msk)  == (UICR_APPROTECT_PALL_Msk))
{
  NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos;
  while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
  NRF_UICR->APPROTECT = (UICR_APPROTECT_PALL_HwDisabled << UICR_APPROTECT_PALL_Pos);
  while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
  NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos;
  while (NRF_NVMC->READY == NVMC_READY_READY_Busy) {}
  NVIC_SystemReset();
}  

NRF_APPROTECT->DISABLE = NRF_UICR->APPROTECT;
```

Note this will be done by the OS soon (a PR is pending) so this is just an example, not a permanent solution.

You can read more about this here (Access port protection controlled by hardware and software):
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fdif.html&cp=5_0_0_3_7_1&anchor=concept_udr_mns_1s

**Comment by alexstanoev** [comment]
@pjonsson brings up a very good point - are you using the nRF52840 dongle or DK?

If you're using the devkit and have connected your USB cable to the JLink port, double-check your connections and look for `1366:0105 SEGGER J-Link` in `lsusb`.

If you're using the dongle you will need the `dfu-upload` target (https://github.com/contiki-ng/contiki-ng/wiki/Platform-nrf52840#compilation-targets) along with setting `PORT=/dev/ttyACMx` where x is your dongle in bootloader mode.

You can find the port your device got allocated by looking at the last lines in `dmesg` after plugging it in - you're looking for Open DFU Bootloader. If not, press the reset button and wait for the red led to start flashing.
Don't forget to apply the udev rules listed in the wiki page or you'll get `LIBUSB_ERROR_ACCESS` back from `nrfutil`.
